### DISCUSSION TITLE
Error reading PDM mic

### DISCUSSION THREAD

**Discussion by JetForMe** [pschatzmann/arduino-audio-tools#1778] [Q&A] [answered]
I cobbled together the code below from the examples and docs. When I run it, I get only this repeated output:

```
[I] StreamCopy.h : 163 - StreamCopy::copy  1024 -> 0 -> 0 bytes - in 0 hops
[E] I2SESP32V1.h : 121 - size_t audio_tools::I2SDriverESP32V1::readBytes(void*, size_t)
```

I have an [Adafruit PDM Mic](https://www.adafruit.com/product/4346) connected to an [Adafruit Matrix Portal S3](https://www.adafruit.com/product/5778) using pins A1 and A3 for clock and data. The WS is hardwired low on this mono microphone.

The code generating the error log is 

```c++
  size_t readBytes(void *dest, size_t size_bytes) {
    size_t result = 0;
    if (i2s_channel_read(rx_chan, dest, size_bytes, &result,
                         ticks_to_wait_read) != ESP_OK) {
      TRACEE();
    }
    return result;
  }
```

I'm not sure what the actual error is, of course. Is it disliking the data it sees? Is it a configuration problem?

```c++
#include "AudioTools.h"

AudioInfo info(44100, 1, 16);
I2SStream i2sStream; // Access I2S as stream
CsvOutput<int16_t> csvStream(Serial);
StreamCopy copier(csvStream, i2sStream); // copy i2sStream to csvStream

const int I2S_SCK = A1;
const int I2S_SD = A3;

// Arduino Setup
void setup(void) {
    Serial.begin(115200);
    AudioToolsLogger.begin(Serial, AudioToolsLogLevel::Info);
    
    auto cfg = i2sStream.defaultConfig(RX_MODE);
    cfg.copyFrom(info);
	cfg.signal_type = PDM;
    cfg.i2s_format = I2S_STD_FORMAT; // or try with I2S_LSB_FORMAT
    cfg.is_master = true;
     // this module nees a master clock if the ESP32 is master
    cfg.use_apll = false;  // try with yes
    cfg.pin_data = I2S_SD; 
    cfg.pin_mck = I2S_SCK; 
    i2sStream.begin(cfg);

    // make sure that we have the correct channels set up
    csvStream.begin(info);

}

// Arduino loop - copy data
void loop() {
    copier.copy();
}
```

**Comment by pschatzmann** [comment, marked as answer]
If you check your log properly you should see the following "Only TX supported for PDM"!
I [just committed the RX support for PDM](https://github.com/pschatzmann/arduino-audio-tools/commit/5214dec10ebaab57d6eb485cb96b4180da476b54): please let me know if this is working....

**Reply by JetForMe** [reply]
Thank you, I'll try that. The logging output was so much, I missed that note entirely.

**Reply by JetForMe** [reply]
Hmm, trying after `git pull` in the `main` branch, the only logging I get is these two lines repeated. I did verify that it is picking up that code.

```
16:03:31.776 -> [E] I2SESP32V1.h : 120 - size_t audio_tools::I2SDriverESP32V1::readBytes(void*, size_t)
16:03:31.776 -> [I] StreamCopy.h : 163 - StreamCopy::copy  1024 -> 0 -> 0 bytes - in 0 hops
```

No other messages.

**Reply by JetForMe** [reply]
I'm surious, if PDM receive wasn't implemented, why is there the `streams-i2s_pdm-serial.ino` example?

**Reply by pschatzmann** [reply]
Espressif changed the I2S functionality!
If you would go back installing  an old ESP32 Arduino 2.0.x Kernal (e.g. 2.0.17) this sketch should still work! 

Fee free to contribute with a correction to make it work again in 3.x...

**Reply by JetForMe** [reply]
Ah, the initial logging was failing because it wasn't ready. I put in a check for that, and now I get this:

```
[I] I2SConfigESP32V1.h : 79 - port_no: 0
[I] I2SConfigESP32V1.h : 80 - is_master: Master
[I] I2SConfigESP32V1.h : 81 - sample rate: 16000
[I] I2SConfigESP32V1.h : 82 - bits per sample: 16
[I] I2SConfigESP32V1.h : 83 - number of channels: 1
[I] I2SConfigESP32V1.h : 84 - signal_type: PDM
[I] I2SConfigESP32V1.h : 85 - buffer_count:6
[I] I2SConfigESP32V1.h : 86 - buffer_size:512
[I] I2SConfigESP32V1.h : 87 - auto_clear: true
[I] I2SConfigESP32V1.h : 98 - pin_mck: 10
[I] I2SConfigESP32V1.h : 100 - pin_bck: 6
[I] I2SConfigESP32V1.h : 102 - pin_ws: 7
[I] I2SConfigESP32V1.h : 104 - pin_data: 3
[E] I2SESP32V1.h : 379 - i2s_channel_init_pdm_rx_mode rx
[E] I2SESP32V1.h : 508 - Channels not started
…returned
Beginning csvStream
…returned
[E] I2SESP32V1.h : 120 - size_t audio_tools::I2SDriverESP32V1::readBytes(void*, size_t)
[I] StreamCopy.h : 163 - StreamCopy::copy  1024 -> 0 -> 0 bytes - in 0 hops
...
```

I'm not sure what "Channels not started" means. Do I need to make another call, or is this the change you're talking about?

My eventual goal is to implement PDM and I2S input for CircuitPython. I started with this library because I thought it would be a good way to verify that I could get my hardware to work properly before taking on the other task.

I'd be happy to try to make the change you suggest to fix things, but I would need more direction. I don't even know how to change to the 2.0.x kernel, as you suggest.

**Comment by pschatzmann** [comment]
Try to activate the ESP32 logging in the Tools menu, this might eventually give another clue.

You can change the installed version in the boards manager on the right: Just type ESP32 to narrow down the list and then for the esp32 from Espressif you can select the version to be used.

The responsible code can be found in src/AudioTools/CoreAudio/AudioI2S/I2SESP32V1.h. As you can see in the log, the logic on line 379 seems to be failing. What ESP version are you using ?

**Reply by pschatzmann** [reply]
Please note that this is [not supported by all processors](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/peripherals/i2s.html#overview-of-all-modes). 

I just realized that you use an S3, so PDM input should work, using a 2.0.x kernel might not be an option because I think the S3 might not be supported there.

**Reply by pschatzmann** [reply]
Activating the Arduino ESP32 Core Debug was giving the hint where the issue was.
I committed [a correction](https://github.com/pschatzmann/arduino-audio-tools/commit/3108e582a6a9a44d3a13716d7127523d9a7f9867) and the initialization should now work properly

**Reply by JetForMe** [reply]
It initializes and returns data now! But the only data I get is `-30935`. I'll have to investigate to make sure I'm seeing data on the pins, etc. Thank you for looking at this!

**Reply by JetForMe** [reply]
Okay, near as I can tell, the pins aren't being set up properly. I'm not sure what I'm doing wrong. On this board, `A1`-`A4` are predefined constants for the pins I want to use. I'm using `A1` for the clock, and `A2` as the data input. These are wired to IO3 (pin 15) and IO9 (pin 17) on the S3. If I toggle `A1` with `digitalWrite(A1, LOW|HIGH)`, I can see it being toggled on the oscilloscope.

If I configure the `I2SStream` with:

```
const int I2S_SCK = A1;
const int I2S_SD = A2;

void setup(void) {
    Serial.begin(115200);
	while (!Serial)
	{
	}
	Serial.println("Hellow");
    AudioToolsLogger.begin(Serial, AudioToolsLogLevel::Info);
    
    auto cfg = i2sStream.defaultConfig(RX_MODE);
    cfg.copyFrom(info);
	cfg.signal_type = PDM;
    cfg.i2s_format = I2S_STD_FORMAT; // or try with I2S_LSB_FORMAT
    cfg.is_master = true;
     // this module nees a master clock if the ESP32 is master
    cfg.use_apll = false;  // try with yes
    cfg.pin_data = I2S_SD; 
    cfg.pin_mck = I2S_SCK;
…
```

I don't see a signal on the clock pin. Am I missing something in my setup?

**Reply by JetForMe** [reply]
I figured it out. The clock pin should be configured on `cfg.pin_bck`. Not sure if you need to update your examples.

But I'm reading samples now! Thank you!

**Comment by pschatzmann** [comment]
The mck is not used for PDM : only pin_data and pin_bck ! 

Please use proper GPIO numbers as documented in the Wiki.
The Ax pins should be used for Analog IO only and in this case you want to use digital pins!
For [the S3 this is by chance the same](https://github.com/espressif/arduino-esp32/blob/master/variants/esp32s3/pins_arduino.h), but this is rather the exception.

I have updated the Wiki and the example
